R/paginate.R

```diff
@@ -12,52 +12,64 @@
 #'
 #' @param url API endpoint
 #' @param limit Number of items per query
+#' @param head Should the query return the data (head = FALSE) or the number of records that would be returned (head = TRUE)?
 #' @param verbose Whether to produce verbose output.
-paginate <- function(url, limit = 10000, verbose = FALSE) {
+paginate <- function(url, head = FALSE, limit = 10000, verbose = FALSE) {
   offset = 0
   results <- list()
 
-  while(TRUE) {
-    url <- url %>%
-      param_set(key = "limit", value = limit) %>%
-      # Prevent converting to scientific notation because "+" has different meaning in URL.
-      param_set(key = "offset", value = format(offset, scientific = FALSE))
+  if (!head) {
+    while(TRUE) {
+      url <- url %>%
+        param_set(key = "limit", value = limit) %>%
+        # Prevent converting to scientific notation because "+" has different meaning in URL.
+        param_set(key = "offset", value = format(offset, scientific = FALSE))
 
-    response <- GET(url)
+      response <- GET(url)
 
-    # Check for "204 NO CONTENT".
-    #
-    if (response$status_code == 204) break
+      # Check for "204 NO CONTENT".
+      #
+      if (response$status_code == 204) break
 
-    check_response_error(response)
-    check_response_json(response)
+      check_response_error(response)
+      check_response_json(response)
 
-    result <- response %>%
-      content(as = "text", encoding = "UTF-8") %>%
-      fromJSON()
+      result <- response %>%
+        content(as = "text", encoding = "UTF-8") %>%
+        fromJSON()
 
-    if (verbose) {
-      message(sprintf("Retrieved %d results.", nrow(result)))
-    }
+      if (verbose) {
+        message(sprintf("Retrieved %d results.", nrow(result)))
+      }
 
-    if (!length(result)) {
-      if (verbose) message("No more rows.")
-      break
-    }
+      if (!length(result)) {
+        if (verbose) message("No more rows.")
+        break
+      }
+
+      results <- c(results, list(result))
 
-    results <- c(results, list(result))
+      if (nrow(result) < limit) {
+        if (verbose) message("Received fewer rows than requested.")
+        break
+      }
 
-    if (nrow(result) < limit) {
-      if (verbose) message("Received fewer rows than requested.")
-      break
+      offset = offset + limit
     }
 
-    offset = offset + limit
+    do.call(rbind, results) %>%
+      as_tibble() %>%
+      mutate_at(vars(any_of(c("brand", "model", "sku"))), as.character) %>%
+      mutate_at(vars(any_of(c("price", "price_promotion"))), as.numeric) %>%
+      mutate_at(vars(any_of(c("time"))), function(time) {as.POSIXct(time, format = "%Y-%m-%dT%H:%M:%S", tz = "UTC")})
+  } else {
+    response <- HEAD(url)
+
+    if (response$status_code == 204) break
```

Comment:
Hi @MeganBeckett, I am wondering about this line. Normally you only have a `break` inside a loop. What was the intended behaviour here? Presumably you want the function to return immediately. What should the return value be? Thanks, Andrew.

Reply:
Mmm, yes, I added this check in as it was already there for the GET request (which is in a `while` loop). As far as I can see, this will also just return immediately, with no value? 

I wasn't sure if there was a reason for the specific check for the status code 204, so included it in the HEAD part of the function too. I can just wrap this in a `while(TRUE)` loop too, or remove?